packages/loopring_v3/contracts/amm/libamm/AmmStatus.sol

```diff
@@ -65,22 +70,34 @@ library AmmStatus
 
             ERC20(token).approve(depositContract, uint(-1));
         }
+
+        // The last token is the pool token
+        S.tokens.push(AmmData.Token({
+            addr: address(this),
+            tokenID: exchange.getTokenID(address(this)),
+            weight: 0 // never used
+        }));
     }
 
     // Anyone is able to shut down the pool when requests aren't being processed any more.
-    function shutdown(
-        AmmData.State storage S,
-        bytes32               txHash
-        )
+    function shutdown(AmmData.State storage S)
         public
     {
-        require(
-            block.timestamp > S.approvedTx[txHash] + AmmData.MAX_AGE_REQUEST_UNTIL_POOL_SHUTDOWN(),
-            "REQUEST_NOT_TOO_OLD"
-        );
+        bytes32 firstExitHash = S.exitLocks[S.exitLocksStartIdx].txHash;
+        uint validUntil = S.approvedTx[firstExitHash];
+        require(validUntil > 0 && validUntil <= block.timestamp, "REQUEST_NOT_TOO_OLD");
+
+        uint size = S.tokens.length;
+
+        // remember the part owned by users collectively
+        for (uint i = 0; i < size; i++) {
+            address token = S.tokens[i].addr;
+            S.withdrawableBeforeShutdown[token] = (token == address(0)) ?
+                address(this).balance :
+                ERC20(token).balanceOf(address(this));
+        }
```

Comment:
Balances withdrawn from `withdrawable` and joints/exits, which we also need to support when shutdown, will need to be subtracted from `withdrawableBeforeShutdown` to keep it correct